File: cpp/0059_Spiral_Matrix_II.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        int top = 0, bottom = n - 1, left = 0, right = n - 1;
        int cnt = 1;
        int d = 1;
        vector<vector<int>> ans(n, vector<int> (n, 0));

        while (bottom >= top && right >= left) {
            if (d == 1) {
                for (int c = left; c <= right; c++) ans[top][c] = cnt++;
                top++;
            } else if (d == 2) {
                for (int r = top; r <= bottom; r++) ans[r][right] = cnt++;
                right--;
            } else if (d == 3) {
                for (int c = right; c >= left; c--) ans[bottom][c] = cnt++;
                bottom--;
            } else {
                for (int r = bottom; r >= top; r--) ans[r][left] = cnt++;
                left++;
            }
            d = (d + 1) % 4;
        }

        return ans;
    }
};

File: python/0059_Spiral_Matrix_II.py
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        top, bottom = 0, n - 1
        left, right = 0, n - 1
        ans = [[0] * n for _ in range(n)]
        cnt = 1
        d = 1

        while bottom >= top and right >= left:
            if d == 1:
                for c in range(left, right + 1):
                    ans[top][c] = cnt
                    cnt += 1
                top += 1
            elif d == 2:
                for r in range(top, bottom + 1):
                    ans[r][right] = cnt
                    cnt += 1
                right -= 1
            elif d == 3:
                for c in range(right, left - 1, -1):
                    ans[bottom][c] = cnt
                    cnt += 1
                bottom -= 1
            else:
                for r in range(bottom, top - 1, -1):
                    ans[r][left] = cnt
                    cnt += 1
                left += 1
            d = (d + 1) % 4
        
        return ans